indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java

```diff
@@ -112,71 +107,66 @@
- * <p>
- * ZK_CLEANUP_TODO : As of 0.11.1, it is required to cleanup task status paths from ZK which are created by the
- * workers to support deprecated RemoteTaskRunner. So a method "scheduleCompletedTaskStatusCleanupFromZk()" is added'
- * which should be removed in the release that removes RemoteTaskRunner legacy ZK updation WorkerTaskMonitor class.
+ * HTTP-based distributed task scheduler that manages assignment of tasks to slots on workers (MiddleManagers).
+ * State information:
+ *  Task state machine is as follows:
+ *    1. PENDING – Task has been submitted to the scheduler.
+ *    2. PENDING_ASSIGN – Task has been assignment to a worker, but has not started running yet.
+ *    3. EXECUTING – Task is running on a worker.
+ *    4. COMPLETE – Task has completed (success/fail).
+ *  Worker state machine is as follows:
+ *    1. READY – Worker is online and ready to receive new tasks.
+ *    2. PENDING_ASSIGN – A task has been submitted to this worker, but has not started running yet.
+ *    3. BLACKLISTED – Worker has too many failed tasks.
+ *    4. LAZY – Worker has no more task running and been marked as reapable by the worker auto-scaler.
  */
 public class HttpRemoteTaskRunner implements WorkerTaskRunner, TaskLogStreamer, WorkerHolder.Listener
 {
-  public static final String TASK_DISCOVERED_COUNT = "task/discovered/count";
-
   private static final EmittingLogger log = new EmittingLogger(HttpRemoteTaskRunner.class);
 
+  public static final String TASK_DISCOVERED_COUNT = "task/discovered/count";
+  private static final int FIND_WORKER_BACKOFF_DELAY_MILLIS = 10_000;
+  private static final int FIND_WORKER_BACKOFF_RETRIES = 4;
+
   private final LifecycleLock lifecycleLock = new LifecycleLock();
 
   // Executor for assigning pending tasks to workers.
   private final ExecutorService pendingTasksExec;
 
   // All known tasks, TaskID -> HttpRemoteTaskRunnerWorkItem
   // This is a ConcurrentMap as some of the reads are done without holding the lock.
-  @GuardedBy("statusLock")
-  private final ConcurrentMap<String, HttpRemoteTaskRunnerWorkItem> tasks = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, HttpRemoteTaskRunnerWorkItem> tasks = new ConcurrentHashMap<>();
 
-  // This is the list of pending tasks in the order they arrived, exclusively manipulated/used by thread that
-  // gives a new task to this class and threads in pendingTasksExec that are responsible for assigning tasks to
-  // workers.
-  @GuardedBy("statusLock")
-  private final List<String> pendingTaskIds = new ArrayList<>();
+  private final PriorityBlockingQueue<PendingTaskQueueItem> pendingTasks = new PriorityBlockingQueue<>();
 
   // All discovered workers, "host:port" -> WorkerHolder
-  private final ConcurrentMap<String, WorkerHolder> workers = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, WorkerHolder> workers = new ConcurrentHashMap<>();
```

Comment:
This should probably be `@GuardedBy("workerStateLock")` since it is always meant to be accessed within that lock.